platform/api/openstack/api.go

```diff
@@ -583,6 +592,106 @@ func (a *API) DeleteImage(imageID string) error {
 	return images.Delete(a.imageClient, imageID).ExtractErr()
 }
 
+func (a *API) PruneKeys(olderThan time.Duration) error {
+	// Build a set of keypair names that are still in use by active servers so
+	// that we don't delete keys that are currently required.
+	usedKeys := make(map[string]struct{})
+
+	srvPager := servers.List(a.computeClient, servers.ListOpts{})
+	srvPages, err := unwrapPages(srvPager, true)
+	if err != nil {
+		return fmt.Errorf("listing servers: %v", err)
+	}
+
+	srvList, err := servers.ExtractServers(srvPages)
+	if err != nil {
+		return fmt.Errorf("extracting servers: %v", err)
+	}
+
+	for _, s := range srvList {
+		if s.KeyName != "" {
+			usedKeys[s.KeyName] = struct{}{}
+		}
+	}
+
+	// List all keypairs in the project.
+	kpPager := keypairs.List(a.computeClient, keypairs.ListOpts{})
+	kpPages, err := unwrapPages(kpPager, true)
+	if err != nil {
+		return fmt.Errorf("listing keypairs: %v", err)
+	}
+
+	kpList, err := keypairs.ExtractKeyPairs(kpPages)
+	if err != nil {
+		return fmt.Errorf("extracting keypairs: %v", err)
+	}
+
+	now := time.Now()
+
+	for _, kp := range kpList {
+		// Skip keypairs that are still in use.
+		if _, inUse := usedKeys[kp.Name]; inUse {
+			continue
+		}
+
+		// Retrieve detailed information in order to obtain the optional
+		// `created_at` field.
+		var detail struct {
+			Keypair struct {
+				CreatedAt string `json:"created_at"`
+			} `json:"keypair"`
+		}
+
+		if err := keypairs.Get(a.computeClient, kp.Name, nil).ExtractInto(&detail); err != nil {
+			// If we fail to obtain details, skip deletion to be safe.
+			plog.Warningf("could not get details for keypair %s: %v", kp.Name, err)
+			continue
+		}
+
+		if detail.Keypair.CreatedAt == "" {
+			// Missing creation timestamp – skip.
+			continue
+		}
+
+		// Attempt to parse the timestamp using a small set of common layouts.
+		var createdTime time.Time
+		var parseErr error
+		for _, layout := range []string{
+			time.RFC3339,
+			"2006-01-02T15:04:05.999999Z07:00", // micro-seconds + TZ
+			"2006-01-02T15:04:05.999999",       // micro-seconds, no TZ
+			"2006-01-02T15:04:05",              // seconds, no TZ
+		} {
+			createdTime, parseErr = time.Parse(layout, detail.Keypair.CreatedAt)
+			if parseErr == nil {
+				// If the layout did not specify timezone information (no "Z07"),
+				// assume the timestamp is in UTC, which is what OpenStack typically
+				// uses internally.
+				if !strings.Contains(layout, "Z07") {
+					createdTime = createdTime.UTC()
+				}
+				break
+			}
+		}
+		if parseErr != nil {
+			plog.Warningf("unable to parse created_at for keypair %s: %v", kp.Name, parseErr)
+			continue
+		}
+
+		if now.Sub(createdTime) > olderThan {
+			if err := a.DeleteKey(kp.Name); err != nil {
+				plog.Warningf("failed deleting stale keypair %s: %v", kp.Name, err)
```

Comment:
The behavior in this case differs a bit from other deletions where it's not a warning but a hard error, yet I think it makes sense to continue and maybe the other calls could also be adjusted to allow continuing - it gives less visibility though and I think that piled up keys are less harmful than piled up running VMs. So, in the end the different behavior for these cases is a good tradeoff?